providers/dns/inwx/inwx.go

```diff
@@ -211,25 +211,27 @@ func (d *DNSProvider) twoFactorAuth(info *goinwx.LoginResponse) error {
 		time.Sleep(sleep)
 	}
 
-	tan, err := totp.GenerateCode(d.config.SharedSecret, time.Now())
+	now := time.Now()
+
+	tan, err := totp.GenerateCode(d.config.SharedSecret, now)
 	if err != nil {
 		return err
 	}
 
-	d.previousUnlock = time.Now()
+	d.previousUnlock = now.Truncate(30 * time.Second)
```

Comment:
Time.Truncate() is rounding down to multiples of the duration since the Go zero time [1], but that is a different baseline time than the Unix epoch on 1970-01-01.

[1] https://pkg.go.dev/time#Time.IsZero

Reply:
Not sure to see what can be difference as we are using second :thinking: 

https://go.dev/play/p/SECZB_iROSw

Maybe I miss something, can you give me more explanation?

Reply:
Based on:
- https://pkg.go.dev/time
- https://github.com/golang/go/issues/15247

leap seconds doesn't seem to be a problem.

Reply:
You are correct, and your solution is more elegant indeed.  Let's use `Time.Truncate()` :)

```
package trunctime                                                                      
                                                                                       
import (                                                                               
        "testing"                                                                      
        "time"                                                                         
)                                                                                      
                                                                                       
func FuzzTruncation(f *testing.F) {                                                    
        for _, uts := range []int64{                                                   
                0, 1234567890, 99,                                                     
        } {                                                                            
                f.Add(uts)                                                             
        }                                                                              
        f.Fuzz(func(t *testing.T, uts int64) {                                         
                if uts <= 0 {                                                          
                        t.Skip("Negative timestamp")                                   
                }                                                                      
                ts := time.Unix(uts, 0)                                                
                                                                                       
                tstrunc := ts.Truncate(30 * time.Second)                               
                utstrunc := time.Unix(30*(uts/30), 0)                                  
                                                                                       
                if !tstrunc.Equal(utstrunc) {                                          
                        t.Errorf("tstrunc %v != utstrunc %v", tstrunc, utstrunc)       
                }                                                                      
        })                                                                             
}
```

I threw a fuzzer at it, to be sure, and it did not find any differences after a few tens-of-million attempts :)

```
...
fuzz: elapsed: 3m36s, execs: 49078748 (227838/sec), new interesting: 2 (total: 5)
fuzz: elapsed: 3m39s, execs: 49767357 (229712/sec), new interesting: 2 (total: 5)
fuzz: elapsed: 3m42s, execs: 50445798 (226109/sec), new interesting: 2 (total: 5)
^Cfuzz: elapsed: 3m43s, execs: 50697296 (229938/sec), new interesting: 2 (total: 5)
PASS
ok  	github.com/gnoack/go-experiment/trunctime	223.097s
```